### PR TITLE
Fixed onLoadChildren being way too slow

### PR DESCRIPTION
**Pull request by lcremmer-ccc**
Hi Adrian,

Here are a couple of fixes to the MediaBrowserService backing up Vanilla's compatibility with MirrorLink.
Clearly at the time I wrote this I didn't fully understand the inner working of Vanilla and my way of retrieving the cover art was really bad.

It is way better now (far less content resolver accesses) but I needed to tweak cover cache a little to make sure a cover can be generated for a song with no path.

Let me know if you have any comments,

Best regards,
Laurent

### CODE REVIEW

**Comment by adrian-bl** [comment]
> It is way better now (far less content resolver accesses) but I needed to tweak cover cache a little 
> to make sure a cover can be generated for a song with no path.

But the MirrorLink device will now show a different cover than the phone itself if the cover loader was active (which is the default)

Did you verify that the content resolver access really was the issue? The code now skips the call to
```
MediaUtils.getSongByTypeId(resolver, mediaType, mediaId);
```
but i would be very surprised if this caused any slowness as the query done by the function should be very quick.

Maybe the slowdown was actually caused by the cover-loader itself? (which will now be skipped as path is null)

**Comment by lcremmer-ccc** [comment]
Hi Adrian,

Good point, the original cause of the slow down was indeed my use of:

Song song = MediaUtils.getSongByTypeId(resolver, mediaType, mediaId);

as getSongByTypeId will itself use a ContentResolver Query to grab the information, and it really started to crawl with 2000+ songs loaded on a device. This is why I replaced it.

But you are right, I've obviously taken a shortcut here, and I should be populating 'path' properly instead of just working around it.

Thanks for the thorough review, much appreciated!

BR-
Laurent

**Comment by adrian-bl** [comment]
Hmm.. did you do any profiling on the function? Because i'd expect it to be quite fast as the query should be trivial on the MediaStore:

https://github.com/vanilla-music/vanilla/blob/master/src/ch/blinkenlights/android/vanilla/MediaUtils.java#L553

(And is used by eg. LazyCover view)

Btw: Is there any way to pass a custom view to the medialink device? If yes, then we could actually re-use the LazyCoverView code with its builtin LRU cache.

**Comment by lcremmer** [comment]
Hi Adrian,

Unfortunately I didn't have the time to do a full profile to figure out if it was building the query or running it that was the root cause, but switching away from getSongTypeById improved load times ten fold in the car. 
Also It seemed like a waste of resource to redo the query for each song where I had (almost) whatever was needed at hand.

The MediaBrowserService doesn't allow cover data to be passed around as a view, it requires a bitmap and unfortunately, doesn't really cater for lazy loading unless you pass the cover data as a URL if you can.

You probably don't see these kind f slowdown since the library UI will only load whichever cover its needs, whereas here I suspect one of the RemoteController having the issue is trying to buffer all of the items (as then onLoadChildren API doesn't really account for partial loading either).

All in all, it is probably just about getting the projection set right from the onset, which I can do.

Thanks for all the helpful input,

Best regards,
Laurent

**Comment by adrian-bl** [comment]
> Also It seemed like a waste of resource to redo the query for each song where I had (almost) 
> whatever was needed at hand.

Yes it is. And the normal cover loader doesn't call it that often as it uses an LRU cache (with the album id as key).

I'll try to do some profiling when i have some spare time (probably in 1-2 months due to work and holidays) but i'd be willing to merge this PR as a 'quickfix' after a simple change.

-> Please remove the change in CoverCache.java: The code really *should* crash if path is null
-> To avoid the crash, just set 'song.path' to some non-existing value. eg:
```
song.path = "/invalid/fixme";
```

**Comment by lcremmer** [comment]
Hi Adrian,

No worries about work and holidays, and no need to rush this either (given I've got seriously sidetracked in the last weeks).

One thing I didn't noticed is that you moved from using a Uri for the Icon to a bitmap to accomodate android wear. Besides getting sluggish on large collections (looks like I can't get songById to return in less than 16ms, even if I had 'limit 1' to the sort clause) it can also put quite the pressure on memory (for a 2000+ song collection :-) ).

One interesting thing is that Wear doesn't seem to be bothered by still using a URI for the current item being played.

With the current state of affairs, maybe the thing to do is to restrict using bitmaps when the dataset being returned isn't too large ?

**Comment by adrian-bl** [comment]
> One thing I didn't noticed is that you moved from using a Uri for the Icon to a bitmap to accomodate 
> android wear. 

Yes, i wanted to 'unify' the cover-loading code and i didn't notice any regression on my Moto360 :-/ (but i do not have a huge test library on my test device).

The reason to move to bitmaps is that we have our own cover loading code which may pick different covers than the one picked up by android. And AFAIK, there is no way to pass these images to MirrorLink using an URI.

>  (looks like I can't get songById to return in less than 16ms, even if I had 'limit 1' to the sort 
clause)

Ouch, that's quite unexpected: I'll have a look at this as improving this would make cover loading faster in the whole app. (Created a reminder for me at https://github.com/vanilla-music/vanilla/issues/433)

So the PR looks fine to me as it is: Shall i merge it?

File: src/ch/blinkenlights/android/vanilla/MirrorLinkMediaBrowserService.java [hunk 1]
```diff
@@ -420,37 +429,60 @@ private String  subtitleForMediaType(int mediaType) {
...
 						.build(),
 						flags);
 				populateMe.add(item);
+				cursor.moveToNext();
 			}
-
 			cursor.close();
 		} catch (Exception e) {
-			Log.d("VanillaMusic","Failed retrieving Media");
+			Log.d("VanillaMusic","Failed retrieving Media " + e.toString());
```
Comment (Adonai) [inline src/ch/blinkenlights/android/vanilla/MirrorLinkMediaBrowserService.java]: Why this? This should be `Log.d("...", "...", e);`. This way you won't lose exception stacktrace.

File: src/ch/blinkenlights/android/vanilla/MirrorLinkMediaBrowserService.java [hunk 2]
```diff
@@ -420,37 +429,60 @@ private String  subtitleForMediaType(int mediaType) {
...
+			cursor.moveToFirst();
+			for (int j = 0; j < count; ++j) {
+				final Long id = cursor.getLong(0);
+				String title = cursor.getString(2);
+				Bitmap cover = null;
+				if(mediaType == MediaUtils.TYPE_SONG) {
+					if(j < 50) {
+						song.populate(cursor);
+						song.flags = 0;
+						if(song.isFilled()) {
```
Comment (Adonai) [inline src/ch/blinkenlights/android/vanilla/MirrorLinkMediaBrowserService.java]: Ternary operation would save you 5 lines: `cover = song.isFilled() ? song.getSmallCover() : sFallbackBitmap`

File: src/ch/blinkenlights/android/vanilla/MirrorLinkMediaBrowserService.java [hunk 3]
```diff
@@ -420,37 +429,60 @@ private String  subtitleForMediaType(int mediaType) {
...
+						Song songForCover = MediaUtils.getSongByTypeId(resolver, mediaType, id);
+						if(songForCover != null && songForCover.isFilled()) {
+							cover = songForCover.getSmallCover(context);
+						}
+						if(cover == null) {
+							cover = sFallbackBitmap;
+						}
+					}
+				} else {
+					cover = CoverBitmap.generatePlaceholderCover(context, 88,88, title);
```
Comment (Adonai) [inline src/ch/blinkenlights/android/vanilla/MirrorLinkMediaBrowserService.java]: Comma not followed by space

File: src/ch/blinkenlights/android/vanilla/MirrorLinkMediaBrowserService.java [hunk 4]
```diff
@@ -420,37 +429,60 @@ private String  subtitleForMediaType(int mediaType) {
...
+			Cursor cursor;
+			if(mediaType == MediaUtils.TYPE_SONG) {
+				cursor = adapter.buildSongQuery(Song.FILLED_PROJECTION).runQuery(resolver);
+			} else {
+				cursor = adapter.query();
 			}
-
+			Song song = new Song(-1);
 			final int flags = (mediaType == MediaUtils.TYPE_SONG || mediaType == MediaUtils.TYPE_PLAYLIST) ? MediaBrowser.MediaItem.FLAG_PLAYABLE : MediaBrowser.MediaItem.FLAG_BROWSABLE;
 			final int count = cursor.getCount();
```
Comment (Adonai) [inline src/ch/blinkenlights/android/vanilla/MirrorLinkMediaBrowserService.java]: Why make variable just to use it in one place?

**Comment by Adonai** [comment]
@lcremmer-ccc , any news on this PR?

**Comment by adrian-bl** [comment]
closing this as it seems to be orphaned.

**Pull request closed without merging**